usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2298 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2020
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+/*
+ * pkcstok_migrate - A tool for migrating token objects to 3.12 format.
+ *
+ */
+
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <errno.h>
+#include <getopt.h>
+#include <memory.h>
+#include <linux/limits.h>
+#include <openssl/evp.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <termios.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <pkcs11types.h>
+
+#include "sw_crypt.h"
+#include "defs.h"
+#include "host_defs.h"
+#include "local_types.h"
+#include "h_extern.h"
+#include "pkcs_utils.h"
+
+
+#define TOKVERSION_00         0x00000000
+#define TOKVERSION_312        0x0003000C
+
+#define HEADER_LEN            64
+#define FOOTER_LEN            16
+
+extern pkcs_trace_level_t trace_level;
+
+/**
+ * Make a 3.12 format OBJECT_PUB:
+ *
+ *   struct OBJECT_PUB {
+ *       //--------------        <--+
+ *       u32 tokversion;            | 16-byte header
+ *       u8 private_flag;           |
+ *       u8 reserved[7];            |
+ *       u32 object_len;            |
+ *       //--------------        <--+
+ *       u8 object[object_len];     | body
+ *       //--------------        <--+
+ *   };
+ */
+CK_RV make_OBJECT_PUB_312(char **obj_new, unsigned int *obj_new_len,
```

Comment:
I guess these make_xxx and decrypt_xxx functions are also somehow available in OCK common code. Maybe not exactly as needed here, but in principle. So we should maybe try to extract these functions from common code and also put them into pkcs_utils.c to be shared between tools and base code.

Reply:
Yes, I already looked into this. The main difference is that the routines in common/loadsave.c use the token data struct, while the migration utility just accesses the files. We may get some reusable code out of this, but I would do that with a later cleanup.